src/main/java/com/emc/pravega/perf/PravegaReaderWorker.java

```diff
@@ -30,60 +28,44 @@
 import io.pravega.client.stream.EventRead;
 import io.pravega.client.stream.ReinitializationRequiredException;
 
-public class PravegaReaderWorker implements Callable<Void> {
+public class PravegaReaderWorker extends ReaderWorker  {
     public final static AtomicInteger eventCount = new AtomicInteger(0);
-    public final long totalEvents;
+
     private final EventStreamReader<String> reader;
-    private final int secondsToRun;
-    private final Instant StartTime;
-    private final int timeout;
-    private final PerfStats stats;
     private final String readerId;
 
     PravegaReaderWorker(int readerId, int secondsToRun, Instant start,
-                        ClientFactory factory, PerfStats stats,
-                        String readergrp, long totalEvents, int timeout) {
-        this.readerId = Integer.toString(readerId);
-        this.secondsToRun = secondsToRun;
-        this.StartTime = start;
-        this.stats = stats;
-        this.totalEvents = totalEvents;
-        this.timeout = timeout;
+                        PerfStats stats, String readergrp, long totalEvents,
+                        int timeout, ClientFactory factory) {
+        super(readerId, secondsToRun, start, stats, readergrp, totalEvents, timeout);
 
+        this.readerId = Integer.toString(readerId);
         reader = factory.createReader(
                 this.readerId, readergrp, new UTF8StringSerializer(), ReaderConfig.builder().build());
     }
 
-    public void cleanupEvents(PerfStats drainStats) throws ReinitializationRequiredException {
-        EventRead<String> event;
-        String ret = null;
-        do {
-            final Instant startTime = Instant.now();
-            event = reader.readNextEvent(timeout);
-            ret = event.getEvent();
-            if (ret != null) {
-                drainStats.recordTime(null, startTime, ret.length());
-            }
-        } while (ret != null);
-        drainStats.printTotal(Instant.now());
-    }
-
     @Override
-    public Void call() throws ReinitializationRequiredException, ExecutionException {
-        EventRead<String> event = null;
-        String ret = null;
+    public  String readData() throws Exception {
         try {
-            do {
-                event = reader.readNextEvent(timeout);
-                ret = event.getEvent();
-                if (ret != null) {
-                    stats.recordTime(null, Instant.ofEpochMilli(Long.parseLong(ret.split(",")[0])), ret.length());
-                }
-            } while ((Duration.between(StartTime, Instant.now()).getSeconds() < secondsToRun) &&
-                    (eventCount.incrementAndGet() < totalEvents));
-        } finally {
-            reader.close();
+            return reader.readNextEvent(timeout).getEvent();
+        } catch(ReinitializationRequiredException e){
+            throw new Exception("ReinitializationRequiredException exception while reading data ");
```

Comment:
This doesn't seem useful.

Reply:
i need your help here Tom,   this what my design in brief.
WriterWorker  - An abstract class which implements call method the performance benchmarking.
                           This class will not know anything about pravega or any other streaming apis (like kafka)

PravegaWriterWorker  - Pravega specific writer class for writer benchmarking. This class extends  "WriterWorker  " .. 

so, here is the problem 

methods of PravegaWriterWorker class can throw pravega specific exceptions but  call method (for performance bench-marking) or any other abstract methods of the base abstract class WriterWorker  will not know about it.  So, here WriterWorker   should be generic.
hence, ReinitializationRequiredException  e is converted as generic exception.

please let me know if you want better of dealing the above scenario.

Reply:
It's fine to convert one exception type to another specific type. But here it's just converting to "Exception" which does not serve any purpose because `ReinitializationRequiredException` already is an Exception. So this rethrowing could be deleted and it wouldn't affect the code in any way besides the string "ReinitializationRequiredException exception while reading data" being printed if the caller logs the exception. 
If you want to create some sort of WorkerFailureException then a conversion might make sense.

Reply:
yes, then in this case ReinitializationRequiredException can be converted to illegalstateExcpetion